packages/docusaurus-logger/src/index.ts

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types */
+
+import pico from 'picocolors';
+
+type InterpolatableValue = string | number | (string | number)[];
+
+const path: import('picocolors/types').Formatter = (msg) =>
+  pico.cyan(pico.underline(msg));
+const id: import('picocolors/types').Formatter = (msg) =>
+  pico.blue(pico.bold(msg));
+const code: import('picocolors/types').Formatter = (msg) =>
+  pico.cyan(`\`${msg}\``);
+const subdue = pico.gray;
+const num = pico.yellow;
+
+function interpolate(
```

Comment:
do we really need this?

I don't like at all `%x` syntax personally, it's impossible to understand easily what it means without reading the doc and you can't even navigate easily to that doc while coding. This makes it harder for external contributors to use/understand the logger.

To me, this is simpler, better, and more explicit: 

```ts
logger.error(`Error while executing ${logger.code("docusaurus build")}`);
```

Reply:
Yes, taking care of external contributors makes sense. However, I don't like embedding functions because Prettier never formats them nicely. What about `` logger.error`Error while executing %code${'docusaurus build'}` ``?

Reply:
what do you think about using native support for interpolation?

```ts
logger.error("Error while executing %s.", logger.code("docusaurus build"));
```

Reply:
@armano2 Sebastien doesn't like sigils like `%s`, personally me neither. It's also very important to have the interpolated value in the _exact_ place where it appears instead of using a placeholder, so we avoid reading code going back and forth. (The reason why we want pipeline operators.)

If you looked at my initial implementation (https://github.com/facebook/docusaurus/pull/5994/commits/b111968b5f41e5def1454963591736e4a6c53bb9) I indeed used placeholders because they were easier to implement. However I'd even prefer `` logger.error(`Error while executing ${logger.code("docusaurus build")}`) `` to `` logger.error("Error while executing %s.", logger.code("docusaurus build")) ``. But at the end of the day I still think template interpolation is the most straightforward and readable.

Reply:
> However, I don't like embedding functions because Prettier never formats them nicely.

It's not very important, we shouldn't care and Prettier has room to improve.

> what do you think about using native support for interpolation?

I'd rather use native JS interpolations (even if `%s` is not very explicit) than having our proprietary interpolation syntax

> It's also very important to have the interpolated value in the exact place where it appears instead of using a placeholder, so we avoid reading code going back and forth. (The reason why we want pipeline operators.)

Yes, but for that the option you don't like works well:

```js
logger.error(`Error while executing ${logger.code("docusaurus build")}`);
```

---

Keep in mind that ultimately, plugins will be injected with this logger, and it will become a public API surface.
The goal is to introduce the smallest API surface that does the job done.
I want to only use "the platform", ie JS native apis that users all know about, and not invent our own syntax/feature for this.

Only if users come with concrete use-cases that really require advanced interpolation features we'll introduce those, but not prematurely.

As far as I see, no usage in our own codebase justify the usage of such interpolation syntax

Reply:
> Keep in mind that ultimately, plugins will be injected with this logger, and it will become a public API surface.

I don't understand that part. Plugins aren't injected with this; `@docusaurus/logger` remains a utility that plugin authors can choose to use or not. It's neither a callback nor a plugin of itself, but just a bunch of utility functions.

Reply:
Ok, let's move on and use this interpolation internally for now

The plan is to inject later a logger inside the plugin constructor, and encourage users to use it so that we can enable/disable plugin logs at a config-level. But this can be designed in a separate	feature and we'll figure out a proper public API surface when we do it later.